hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java

```diff
@@ -0,0 +1,421 @@
+                            ManagedRocksDB db,
+                            List<ContainerKeyInfo> containerKeyInfos,
+                            String tableName)
+      throws IOException {
+    long keysProcessed = 0;
+    DBColumnFamilyDefinition<?, ?> columnFamilyDefinition =
+        dbDefinition.getColumnFamily(tableName);
+    if (columnFamilyDefinition == null) {
+      throw new IllegalStateException(
+          "Table with name" + tableName + " not found");
+    }
+
+    ColumnFamilyHandle columnFamilyHandle = getColumnFamilyHandle(
+        columnFamilyDefinition.getName().getBytes(UTF_8),
+        columnFamilyHandles);
+    if (columnFamilyHandle == null) {
+      throw new IllegalStateException("columnFamilyHandle is null");
+    }
+
+    try (ManagedRocksIterator iterator = new ManagedRocksIterator(
+        db.get().newIterator(columnFamilyHandle))) {
+      iterator.get().seekToFirst();
+      while (iterator.get().isValid()) {
+        OmKeyInfo value = ((OmKeyInfo) columnFamilyDefinition.getValueCodec()
+            .fromPersistedFormat(iterator.get().value()));
+        List<OmKeyLocationInfoGroup> keyLocationVersions =
+            value.getKeyLocationVersions();
+        if (Objects.isNull(keyLocationVersions)) {
+          iterator.get().next();
+          keysProcessed++;
+          continue;
+        }
+
+        long volumeId = 0;
+        long bucketId = 0;
+        // volumeId and bucketId are only applicable to file table
+        if (tableName.equals(FILE_TABLE)) {
+          String key = new String(iterator.get().key(), UTF_8);
+          String[] keyParts = key.split(OM_KEY_PREFIX);
+          volumeId = Long.parseLong(keyParts[1]);
+          bucketId = Long.parseLong(keyParts[2]);
+        }
+
+        for (OmKeyLocationInfoGroup locationInfoGroup : keyLocationVersions) {
+          for (List<OmKeyLocationInfo> locationInfos :
+              locationInfoGroup.getLocationVersionMap().values()) {
+            for (OmKeyLocationInfo locationInfo : locationInfos) {
+              if (containerIds.contains(locationInfo.getContainerID())) {
+                // Generate asbolute key path for FSO keys
+                StringBuilder keyName = new StringBuilder();
+                if (tableName.equals(FILE_TABLE)) {
+                  // Load directory table only after the first fso key is found
+                  // to reduce necessary load if there are not fso keys
+                  if (!isDirTableLoaded) {
+                    long start = System.currentTimeMillis();
+                    directoryTable = getDirectoryTableData(parent.getDbPath());
+                    long end = System.currentTimeMillis();
+                    out().println(
+                        "directoryTable loaded in " + (end - start) + " ms.");
+                    isDirTableLoaded = true;
+                  }
+                  keyName.append(getFsoKeyPrefix(volumeId, bucketId, value));
```

Comment:
Looks like this is doing a full directory walk for every key found. I think this approach would be better:
1. Scan the file table and save all the matching keys.
2. If no keys in the file table match, we are done.
3. Else, do one walk through the directory table to build the full paths for all of those keys.
As it is this tool is storing all matches in memory either way. 

A more robust approach would be to have the tool print json matches as it goes so that it works with any size input and output. Then we would walk the directory and file table in one iteration to check for matches.

Reply:
There seems to be multiple approaches on how to print output. There is something like `ListVolumeHandler` for listing volumes which prints out proper JSON format and then there is something like `ListSubcommand` for listing containers which prints out JSON-like format but not true JSON. In order to have it printed out as true JSON format, data for printing needs to be kept in memory and accessed at the time of printing out.

With above in mind would you like to see proper JSON format which is better formatted and easier for end user to read or JSON-like format which is faster performance-wise but uglier for end user?